### ISSUE TITLE
TCP Metrics not being recorded accurately

### ISSUE THREAD

**Issue by whitleykeith** [istio/istio#23991]
**Bug description**
When installing the bookinfo sample + MongoDB, we are seeing an issue where TCP Traffic is inaccurately being recorded. 

For traffic calling to that destination, we see the destination service being $SERVICE.$NAMESPACE.svc.cluster.local with unknown destination workload/namespace.

When the source side of that traffic is being recorded, we see unknowns on everything.

**Expected behavior**

I expect to see TCP traffic recorded accurately

**Steps to reproduce the bug**

Install bookinfo sample with mongodb and generate requests

Image:

![Screen Shot 2020-05-20 at 11 08 54 AM](https://user-images.githubusercontent.com/21354274/82463035-5296cb80-9a8a-11ea-957e-0b56c73e1462.png)

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

Istio 1.5.1
EKS v1.15.11-eks-af3caf

**How was Istio installed?**
Standalone operator

**Environment where bug was observed (cloud vendor, OS, etc)**
EKS and Docker-Desktop

**Comment by whitleykeith** [comment]
Update as I realized a similar issue was closed: #23963

While turning mTLS on resolved the issues with bookinfo/mongo, I still see this issue occuring in other TCP services in our cluster like Zookeeper/Kafka. I think it occurs mainly in TCP traffic with Statefulsets, although it happens even with their client services.

**Comment by jshaughn** [comment]
@whitleykeith Which version of Kiali are you using?  We enhanced our handling of "headless" telemetry like Kafka I think in Kiali 1.17, you may want to give that a try and see if it helps.

**Comment by whitleykeith** [comment]
We're on Kiali 1.18, but I can validate that the envoy metrics themselves are inaccurate.

**Comment by bianpengyuan** [comment]
@whitleykeith could you paste here an example problematic Istio TCP metrics about the headless service? It would be great if you could provide both client and server side metric. I saw that you pasted zookeeper sidecar config dump, it would be nice to also get client side core dump as well, be cause the metadata exchange involves both side.

**Comment by whitleykeith** [comment]
So these are the metrics from the sidecar.

Since Zookeeper tries to talk to itself to establish a quorum you should see both client and server side metrics here. 

reporter: Source

```
istio_tcp_connections_opened_total{reporter="source",source_workload="zookeeper-news",source_workload_namespace="2269-search-clusters",source_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",source_app="zookeeper-news",source_version="unknown",destination_workload="unknown",destination_workload_namespace="unknown",destination_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",destination_app="unknown",destination_version="unknown",destination_service="zookeeper-news-headless.2269-search-clusters.svc.cluster.local",destination_service_name="zookeeper-news-headless.2269-search-clusters.svc.cluster.local",destination_service_namespace="unknown",request_protocol="tcp",response_flags="-",connection_security_policy="unknown",source_canonical_service="zookeeper-news",destination_canonical_service="unknown",source_canonical_revision="latest",destination_canonical_revision="latest"} 1
# TYPE istio_tcp_received_bytes_total counter
istio_tcp_received_bytes_total{reporter="source",source_workload="zookeeper-news",source_workload_namespace="2269-search-clusters",source_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",source_app="zookeeper-news",source_version="unknown",destination_workload="unknown",destination_workload_namespace="unknown",destination_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",destination_app="unknown",destination_version="unknown",destination_service="zookeeper-news-headless.2269-search-clusters.svc.cluster.local",destination_service_name="zookeeper-news-headless.2269-search-clusters.svc.cluster.local",destination_service_namespace="unknown",request_protocol="tcp",response_flags="-",connection_security_policy="unknown",source_canonical_service="zookeeper-news",destination_canonical_service="unknown",source_canonical_revision="latest",destination_canonical_revision="latest"} 7175693
# TYPE istio_tcp_connections_closed_total counter
istio_tcp_connections_closed_total{reporter="source",source_workload="zookeeper-news",source_workload_namespace="2269-search-clusters",source_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",source_app="zookeeper-news",source_version="unknown",destination_workload="unknown",destination_workload_namespace="unknown",destination_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",destination_app="unknown",destination_version="unknown",destination_service="zookeeper-news-headless.2269-search-clusters.svc.cluster.local",destination_service_name="zookeeper-news-headless.2269-search-clusters.svc.cluster.local",destination_service_namespace="unknown",request_protocol="tcp",response_flags="-",connection_security_policy="unknown",source_canonical_service="zookeeper-news",destination_canonical_service="unknown",source_canonical_revision="latest",destination_canonical_revision="latest"} 0
# TYPE istio_tcp_sent_bytes_total counter
istio_tcp_sent_bytes_total{reporter="source",source_workload="zookeeper-news",source_workload_namespace="2269-search-clusters",source_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",source_app="zookeeper-news",source_version="unknown",destination_workload="unknown",destination_workload_namespace="unknown",destination_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",destination_app="unknown",destination_version="unknown",destination_service="zookeeper-news-headless.2269-search-clusters.svc.cluster.local",destination_service_name="zookeeper-news-headless.2269-search-clusters.svc.cluster.local",destination_service_namespace="unknown",request_protocol="tcp",response_flags="-",connection_security_policy="unknown",source_canonical_service="zookeeper-news",destination_canonical_service="unknown",source_canonical_revision="latest",destination_canonical_revision="latest"} 4654615
```

reporter: destination
```
istio_tcp_connections_opened_total{reporter="destination",source_workload="unknown",source_workload_namespace="unknown",source_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",source_app="unknown",source_version="unknown",destination_workload="zookeeper-news",destination_workload_namespace="2269-search-clusters",destination_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",destination_app="zookeeper-news",destination_version="unknown",destination_service="zookeeper-news-client.2269-search-clusters.svc.cluster.local",destination_service_name="zookeeper-news-client",destination_service_namespace="2269-search-clusters",request_protocol="tcp",response_flags="-",connection_security_policy="mutual_tls",source_canonical_service="unknown",destination_canonical_service="zookeeper-news",source_canonical_revision="latest",destination_canonical_revision="latest"} 8
istio_tcp_connections_closed_total{reporter="destination",source_workload="unknown",source_workload_namespace="unknown",source_principal="unknown",source_app="unknown",source_version="unknown",destination_workload="zookeeper-news",destination_workload_namespace="2269-search-clusters",destination_principal="unknown",destination_app="zookeeper-news",destination_version="unknown",destination_service="zookeeper-news-client.2269-search-clusters.svc.cluster.local",destination_service_name="zookeeper-news-client",destination_service_namespace="2269-search-clusters",request_protocol="tcp",response_flags="-",connection_security_policy="none",source_canonical_service="unknown",destination_canonical_service="zookeeper-news",source_canonical_revision="latest",destination_canonical_revision="latest"} 125
istio_tcp_connections_opened_total{reporter="destination",source_workload="unknown",source_workload_namespace="unknown",source_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",source_app="unknown",source_version="unknown",destination_workload="zookeeper-news",destination_workload_namespace="2269-search-clusters",destination_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",destination_app="zookeeper-news",destination_version="unknown",destination_service="zookeeper-news-headless.2269-search-clusters.svc.cluster.local",destination_service_name="zookeeper-news-headless",destination_service_namespace="2269-search-clusters",request_protocol="tcp",response_flags="-",connection_security_policy="mutual_tls",source_canonical_service="unknown",destination_canonical_service="zookeeper-news",source_canonical_revision="latest",destination_canonical_revision="latest"} 4
istio_tcp_connections_closed_total{reporter="destination",source_workload="unknown",source_workload_namespace="unknown",source_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",source_app="unknown",source_version="unknown",destination_workload="zookeeper-news",destination_workload_namespace="2269-search-clusters",destination_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",destination_app="zookeeper-news",destination_version="unknown",destination_service="zookeeper-news-headless.2269-search-clusters.svc.cluster.local",destination_service_name="zookeeper-news-headless",destination_service_namespace="2269-search-clusters",request_protocol="tcp",response_flags="-",connection_security_policy="mutual_tls",source_canonical_service="unknown",destination_canonical_service="zookeeper-news",source_canonical_revision="latest",destination_canonical_revision="latest"} 1
istio_tcp_connections_opened_total{reporter="destination",source_workload="unknown",source_workload_namespace="unknown",source_principal="unknown",source_app="unknown",source_version="unknown",destination_workload="zookeeper-news",destination_workload_namespace="2269-search-clusters",destination_principal="unknown",destination_app="zookeeper-news",destination_version="unknown",destination_service="zookeeper-news-client.2269-search-clusters.svc.cluster.local",destination_service_name="zookeeper-news-client",destination_service_namespace="2269-search-clusters",request_protocol="tcp",response_flags="-",connection_security_policy="none",source_canonical_service="unknown",destination_canonical_service="zookeeper-news",source_canonical_revision="latest",destination_canonical_revision="latest"} 127
istio_tcp_connections_closed_total{reporter="destination",source_workload="unknown",source_workload_namespace="unknown",source_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",source_app="unknown",source_version="unknown",destination_workload="zookeeper-news",destination_workload_namespace="2269-search-clusters",destination_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",destination_app="zookeeper-news",destination_version="unknown",destination_service="zookeeper-news-client.2269-search-clusters.svc.cluster.local",destination_service_name="zookeeper-news-client",destination_service_namespace="2269-search-clusters",request_protocol="tcp",response_flags="-",connection_security_policy="mutual_tls",source_canonical_service="unknown",destination_canonical_service="zookeeper-news",source_canonical_revision="latest",destination_canonical_revision="latest"} 8
istio_tcp_received_bytes_total{reporter="destination",source_workload="unknown",source_workload_namespace="unknown",source_principal="unknown",source_app="unknown",source_version="unknown",destination_workload="zookeeper-news",destination_workload_namespace="2269-search-clusters",destination_principal="unknown",destination_app="zookeeper-news",destination_version="unknown",destination_service="zookeeper-news-client.2269-search-clusters.svc.cluster.local",destination_service_name="zookeeper-news-client",destination_service_namespace="2269-search-clusters",request_protocol="tcp",response_flags="-",connection_security_policy="none",source_canonical_service="unknown",destination_canonical_service="zookeeper-news",source_canonical_revision="latest",destination_canonical_revision="latest"} 9243739
istio_tcp_received_bytes_total{reporter="destination",source_workload="unknown",source_workload_namespace="unknown",source_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",source_app="unknown",source_version="unknown",destination_workload="zookeeper-news",destination_workload_namespace="2269-search-clusters",destination_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",destination_app="zookeeper-news",destination_version="unknown",destination_service="zookeeper-news-headless.2269-search-clusters.svc.cluster.local",destination_service_name="zookeeper-news-headless",destination_service_namespace="2269-search-clusters",request_protocol="tcp",response_flags="-",connection_security_policy="mutual_tls",source_canonical_service="unknown",destination_canonical_service="zookeeper-news",source_canonical_revision="latest",destination_canonical_revision="latest"} 29871
istio_tcp_sent_bytes_total{reporter="destination",source_workload="unknown",source_workload_namespace="unknown",source_principal="unknown",source_app="unknown",source_version="unknown",destination_workload="zookeeper-news",destination_workload_namespace="2269-search-clusters",destination_principal="unknown",destination_app="zookeeper-news",destination_version="unknown",destination_service="zookeeper-news-client.2269-search-clusters.svc.cluster.local",destination_service_name="zookeeper-news-client",destination_service_namespace="2269-search-clusters",request_protocol="tcp",response_flags="-",connection_security_policy="none",source_canonical_service="unknown",destination_canonical_service="zookeeper-news",source_canonical_revision="latest",destination_canonical_revision="latest"} 90147313
istio_tcp_sent_bytes_total{reporter="destination",source_workload="unknown",source_workload_namespace="unknown",source_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",source_app="unknown",source_version="unknown",destination_workload="zookeeper-news",destination_workload_namespace="2269-search-clusters",destination_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",destination_app="zookeeper-news",destination_version="unknown",destination_service="zookeeper-news-headless.2269-search-clusters.svc.cluster.local",destination_service_name="zookeeper-news-headless",destination_service_namespace="2269-search-clusters",request_protocol="tcp",response_flags="-",connection_security_policy="mutual_tls",source_canonical_service="unknown",destination_canonical_service="zookeeper-news",source_canonical_revision="latest",destination_canonical_revision="latest"} 30825
istio_tcp_received_bytes_total{reporter="destination",source_workload="unknown",source_workload_namespace="unknown",source_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",source_app="unknown",source_version="unknown",destination_workload="zookeeper-news",destination_workload_namespace="2269-search-clusters",destination_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",destination_app="zookeeper-news",destination_version="unknown",destination_service="zookeeper-news-client.2269-search-clusters.svc.cluster.local",destination_service_name="zookeeper-news-client",destination_service_namespace="2269-search-clusters",request_protocol="tcp",response_flags="-",connection_security_policy="mutual_tls",source_canonical_service="unknown",destination_canonical_service="zookeeper-news",source_canonical_revision="latest",destination_canonical_revision="latest"} 392
istio_tcp_sent_bytes_total{reporter="destination",source_workload="unknown",source_workload_namespace="unknown",source_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",source_app="unknown",source_version="unknown",destination_workload="zookeeper-news",destination_workload_namespace="2269-search-clusters",destination_principal="spiffe://cluster.local/ns/2269-search-clusters/sa/default",destination_app="zookeeper-news",destination_version="unknown",destination_service="zookeeper-news-client.2269-search-clusters.svc.cluster.local",destination_service_name="zookeeper-news-client",destination_service_namespace="2269-search-clusters",request_protocol="tcp",response_flags="-",connection_security_policy="mutual_tls",source_canonical_service="unknown",destination_canonical_service="zookeeper-news",source_canonical_revision="latest",destination_canonical_revision="latest"} 0
```

What part of the config dump for ZKs sidecar? don't want to paste a 20k line json file that makes it hard for you guys to sift through.

**Comment by bianpengyuan** [comment]
@whitleykeith The whole dump would be great if you are OK with it.

**Comment by whitleykeith** [comment]
The ZK config dump was actually uploaded on this issue: https://github.com/istio/istio/issues/23963#issuecomment-632715549

**Comment by bianpengyuan** [comment]
To confirm, you are using auto mtls and zookeeper peer communication uses port 2181, right?

If so, this actually works as expected, headless service does not support auto mtls now and you'll have to use destination rule to enforce client to use mtls. See 
https://github.com/istio/istio/issues/21964 for more detail.

**Comment by whitleykeith** [comment]
Didn't seem to change anything, however, I think the issue is that the zookeeper operator we use creates the service with bad port names (i.e. not prefixed with tcp). We were able to fix it by prefixing the ports

**Comment by bianpengyuan** [comment]
Ok, thanks! TCP metadata exchange indeed is broken in 1.6.0 / 1.5.4 with proto sniffing. Should be fixed in 1.6.1 and 1.5.5: https://github.com/istio/istio/issues/23907. Naming port properly would also do the job.
